pkg/splunk/enterprise/workerpool.go

```diff
@@ -0,0 +1,191 @@
+
+		object := getRemoteObjectFromS3Response(job.AppName, w.S3Reponse)
+		localFile := getLocalAppFileName(w.DownloadPath, job.AppName, *object.Etag)
+		remoteFile := *object.Key
+
+		// download the app from remote storage
+		err := w.S3ClientMgr.DownloadApp(remoteFile, localFile, *object.Etag)
+		if err != nil {
+			scopedLog.Error(err, "unable to download app", "appName", job.AppName)
+			setAppDownloadState(job, enterpriseApi.DownloadError)
+		} else {
+			// download is successfull, update the state
+			setAppDownloadState(job, enterpriseApi.DownloadComplete)
+
+			//TODO: gaurav, need to fill in below code
+			// copy the app to the splunk pod
+
+			// untar it, if it is cluster scoped
+
+			// if everything is successful, delete the app from operator pod
+		}
+
+		scopedLog.Info("finished job")
+		w.mu.Lock()
+		w.availableDiskSpace = w.availableDiskSpace + job.Size
+		scopedLog.Info("availableDiskSpace after freeing up space", "availableDiskSpace", w.availableDiskSpace)
+		w.mu.Unlock()
+
+		w.wg.Done()
+	}
+}
+
+// isAppAlreadyDownloaded checks if the app is already downloaded on the operator pod by checking a couple of things -
+// 1. is the app bundle present on the operator pod &&
+// 2. if the size on disk matches the size reported from remote storage
+func (w *AppInstallWorkerPool) isAppAlreadyDownloaded(app *enterpriseApi.AppDeploymentInfo) bool {
+	scopedLog := log.WithName("isAppAlreadyDownloaded").WithValues("app name", app.AppName)
+	object := getRemoteObjectFromS3Response(app.AppName, w.S3Reponse)
+	localAppFileName := getLocalAppFileName(w.DownloadPath, app.AppName, *object.Etag)
+
+	var fileInfo os.FileInfo
+	var err error
+
+	// 1. First check if the app is present on operator pod
+	file, err := os.Open(localAppFileName)
+
+	if os.IsNotExist(err) {
+		scopedLog.Info("App not present on operator pod")
+		return false
+	}
+	defer file.Close()
+
+	// 2. Now, since the app is present, check the size
+	fileInfo, _ = os.Stat(localAppFileName)
+	localSize := fileInfo.Size()
+	scopedLog.Info("Size of app on operator pod", "size", localSize)
+
+	remoteSize := *object.Size
+	if localSize != remoteSize {
+		err = fmt.Errorf("size of app on operator pod does not match size on remote storage. localSize: %d, remoteSize:%d", localSize, remoteSize)
+		scopedLog.Error(err, "need to re-download app")
+		return false
+	}
+
+	return true
+}
+
+// Start kicks off the workers to do the job of app installation
+func (w *AppInstallWorkerPool) Start() {
+	// start the workers
+	for i := 1; i <= int(w.NumOfWorkers); i++ {
+		go w.Run(i, w.JobsChan)
+	}
+
+	var jobsAdded int
+	// add the apps to the jobs channel so that workers can pick them up
+	for {
+		for index := 0; index < len(w.JobList); index++ {
+			app := w.JobList[index]
+
+			switch getAppDownloadState(app) {
+			case enterpriseApi.DownloadNotStarted:
+				// check if we have already downloaded the app
+				if w.isAppAlreadyDownloaded(app) {
```

Comment:
Under what circumstances would a successfully downloaded app have status as `DownloadNotStarted`?

Reply:
Under rare scenario, where while we have downloaded the app but have not yet set the download state as DownloadComplete and the operator crashed/restarted. Then we again come here to check if we had already downloaded the app, if yes then we don't need to redownload again.